### PR TITLE
BattleBots

### PR DESCRIPTION
**Pull request by Kittnz**

### CODE REVIEW

File: src/game/PlayerBots/BattleBotWaypoints.cpp [hunk 1]
```diff
@@ -202,23 +213,25 @@ void BattleBotWaypoints::StartMove(Player* me, float waypoint_x, float waypoint_
...
     me->GetMotionMaster()->MovePoint(0, waypoint_x, waypoint_y, waypoint_z, MOVE_PATHFINDING);
 }
 
 bool BattleBotWaypoints::ReachedWaypoint(Player* me)
 {
-    if (3 > sqrt(((me->GetPositionX() - currentWaypointX) * (me->GetPositionX() - currentWaypointX)) + ((me->GetPositionY() - currentWaypointY) * (me->GetPositionY() - currentWaypointY)) + ((me->GetPositionZ() - currentWaypointZ) * (me->GetPositionZ() - currentWaypointZ))))
+    if (!me->IsMoving())
+        return false;
+
+    if (2 > sqrt(((me->GetPositionX() - currentWaypointX) * (me->GetPositionX() - currentWaypointX)) + ((me->GetPositionY() - currentWaypointY) * (me->GetPositionY() - currentWaypointY)) + ((me->GetPositionZ() - currentWaypointZ) * (me->GetPositionZ() - currentWaypointZ))))
```
Comment (ratkosrb) [inline src/game/PlayerBots/BattleBotWaypoints.cpp]: There is a GetDistance function for this.

File: src/game/PlayerBots/BattleBotAI.cpp [hunk 2]
```diff
@@ -885,6 +889,23 @@ void BattleBotAI::UpdateAI(uint32 const diff)
...
+        {
+            if (itr.second.requiredClass == me->GetClass())
+            {
+                if (m_hasGear == false)
+                {
+                    std::string str = std::to_string(itr.first);
+                    char* cstr = new char[str.length() + 1];
+                    strcpy(cstr, str.c_str());
+
+                    ChatHandler(me).HandleCharacterPremadeGearCommand(cstr);
```
Comment (ratkosrb) [inline src/game/PlayerBots/BattleBotAI.cpp]: You do not need to invoke chat commands here. Just call the ObjectMgr function to apply template.

File: src/game/PlayerBots/BattleBotAI.cpp [hunk 3]
```diff
@@ -278,6 +278,70 @@ void BattleBotAI::PopulateSpellData()
...
+                        m_spells.mage.pFireball->Id < pSpellEntry->Id)
+                        m_spells.mage.pFireball = pSpellEntry;
+                }
+                else if (pSpellEntry->SpellName[0].find("Arcane Explosion") != std::string::npos)
+                {
+                    if (!m_spells.mage.pArcaneExplosion ||
+                        m_spells.mage.pArcaneExplosion->Id < pSpellEntry->Id)
+                        m_spells.mage.pArcaneExplosion = pSpellEntry;
+                }
+                else if (pSpellEntry->SpellName[0].find("FrostNova") != std::string::npos)
```
Comment (ratkosrb) [inline src/game/PlayerBots/BattleBotAI.cpp]: missing space

File: src/game/Group/Group.cpp [hunk 4]
```diff
@@ -1421,7 +1421,7 @@ bool Group::_addMember(ObjectGuid guid, char const* name, bool isAssistant, uint
                 player->m_InstanceValid = true;
     }
 
-    if (!isBGGroup())
+    if (!isBGGroup() && !(player && player->GetSession()->GetBot()))
```
Comment (Kittnz) [inline src/game/Group/Group.cpp:1424]: why is this needed?

Comment (ratkosrb) [reply src/game/Group/Group.cpp:1424]: To prevent bot groups from being saved in database. Otherwise you get a whole bunch of startup errors next time about non-existent character in group.
